### DISCUSSION TITLE
Keycloak JavaScript adapter, Authorization Code flow

### DISCUSSION THREAD

**Discussion by borodjuka** [keycloak/keycloak#34705] [Q&A] [answered]
In documentation of your JavaScript adapter, in section Implicit and hybrid flow, says:
_By default, the adapter uses the [Authorization Code](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth) flow._

_With this flow, the Keycloak server returns an authorization code, not an authentication token, to the application. The JavaScript adapter exchanges the code for an access token and a refresh token after the browser is redirected back to the application._

But in referenced specification stands that tokens shouldn't be revealed to User Agent (browser).

Property | Authorization Code Flow | Implicit Flow | Hybrid Flow
-- | -- | -- | --
All tokens returned from Authorization Endpoint | no | yes | no
All tokens returned from Token Endpoint | yes | no | no
**Tokens not revealed to User Agent**| **yes**| no | no
Client can be authenticated | yes | no | yes
Refresh Token possible | yes | no | yes
Communication in one round trip | no | yes | no
Most communication server-to-server | yes | no | varies

 If I understand protocol well, token should be exchanged via back channel.  
 
 This is clear from [OAuth 2.0 RFC, section 4](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1)
 
     +----------+
     | Resource |
     |   Owner  |
     |          |
     +----------+
          ^
          |
         (B)
     +----|-----+          Client Identifier      +---------------+
     |         -+----(A)-- & Redirection URI ---->|               |
     |  User-   |                                 | Authorization |
     |  Agent  -+----(B)-- User authenticates --->|     Server    |
     |          |                                 |               |
     |         -+----(C)-- Authorization Code ---<|               |
     +-|----|---+                                 +---------------+
       |    |                                         ^      v
      (A)  (C)                                        |      |
       |    |                                         |      |
       ^    v                                         |      |
     +---------+                                      |      |
     |         |>---(D)-- Authorization Code ---------'      |
     |  Client |          & Redirection URI                  |
     |         |                                             |
     |         |<---(E)----- Access Token -------------------'
     +---------+       (w/ Optional Refresh Token)

Why then hustle with codes, just request tokens strait away? I'm not sure I'm just confused.

This is similar question (https://github.com/keycloak/keycloak/discussions/21103) product of same confusion i think.

**Comment by embesozzi** [comment, marked as answer]
Generally speaking, when you are implementing the Authorization Code flow, it is typically used in a web app with a confidential client (because you have a backend) or the Authorization Code + PKCE flow in an SPA or native app with a public client. The app receives the authorization code and then negotiates the token. 
If you have a backend (depending on the type of app), some information may or may not be exposed in the browser. In the SPA scenario, a proxy or Backend for Frontend (BFF) [1] can help if you want to avoid exposing sensitive information in the browser.

The Implicit flow is no longer recommended; therefore, SPAs must use the Authorization Code + PKCE flow.

Following the standard and best practices, this is the correct way to implement it.

[1] https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps-19
